common/pulp_docker/common/models.py

```diff
@@ -170,6 +173,48 @@ def __init__(self, digest, name, tag, architecture, fs_layers, history, schema_v
                 "The DockerManifest class only supports Docker v2, Schema 1 manifests.")
         self.schema_version = schema_version
 
+    @staticmethod
+    def digest(manifest):
+        """
+        Calculate the sha256 digest of the Manifest, given in JSON.
+
+        :param manifest: The raw JSON representation of the Manifest.
+        :type  manifest: basestring
+        :return:         The digest of the given Manifest
+        :rtype:          basestring
+        """
+        decoded_manifest = json.loads(manifest, encoding='utf-8')
+        if 'signatures' in decoded_manifest:
+            # This manifest contains signatures. Unfortunately, the Docker manifest digest
+            # is calculated on the unsigned version of the Manifest so we need to remove the
+            # signatures. To do this, we will look at the 'protected' key within the first
+            # signature. This key indexes a (malformed) base64 encoded JSON dictionary that
+            # tells us how many bytes of the manifest we need to keep before the signature
+            # appears in the original JSON. We will strip out the bytes after this cutoff point,
+            # add back a trailing newline and closing curly brace, and then calculate the sha256
+            # sum of the transformed JSON to get the digest.
+            protected = decoded_manifest['signatures'][0]['protected']
+            # Docker has not included the required padding at the end of the base64 encoded
+            # 'protected' block. We need to add the correct number of ='s signs back.
+            protected = protected.replace('\n', '').replace(' ', '')
+            # It is illegal base64 for the remainder to be 1 when the length of the block is
+            # divided by 4.
+            if len(protected) % 4 == 1:
+                raise ValueError(
+                    _('Invalid base64 "protected" block in the first manifest signature.'))
+            # Add back the missing padding characters, based on the length of the encoded string
+            paddings = {0: '', 2: '==', 3: '='}
+            protected += paddings[len(protected) % 4]
+            # Now let's decode the base64 and load it as a dictionary so we can get the length
+            protected = base64.b64decode(protected)
+            protected = json.loads(protected)
+            # This is the length of the signed portion of the Manifest, except for a trailing
+            # newline and closing curly brace.
+            signed_length = protected['formatLength']
+            # Now we can reconstruct the original Manifest that the digest should be based on.
+            manifest = manifest[:signed_length] + '\n}'
```

Comment:
This is likely to work, but it seems more robust to use the ``formatTail`` value from the ``protected`` blob. It is safest for us to assume they put it there because it might change. Docker's own code uses the ``formatTail`` value.